### PR TITLE
Select funding to apply for

### PR DESCRIPTION
**Pull request by marctemp**
https://eaflood.atlassian.net/browse/SFI-241

### CODE REVIEW

File: app/routes/funding/models/what-funding.js [hunk 1]
```diff
@@ -1,3 +1,5 @@
+const { mapStandards } = require('../../../what-funding')
```
Comment (marctemp) [inline app/routes/funding/models/what-funding.js:1]: Attempted to separate responsibility between files.

My understanding is the model should only have information received from the route and produce an object used by the view, hence I have abstracted the functionality of deciding which standards are displayed into a same named JS module.

Is this a valid approach?

File: app/routes/funding/what-funding.js [hunk 2]
```diff
@@ -24,35 +27,37 @@ module.exports = [{
   options: {
     validate: {
       payload: Joi.object({
-        standard: Joi.array().items(Joi.string()).single()
+        standard: Joi.required()
```
Comment (marctemp) [inline app/routes/funding/what-funding.js:30]: Joi did not respond well to attempting to validate the standard payload was an array with at least 1 item in it, therefore I used required. Does mean a user could POST a payload other than an array to this endpoint most likely disrupting the code, but they would need authorisation token to do so.

File: app/routes/funding/what-funding.js [hunk 3]
```diff
@@ -24,35 +27,37 @@ module.exports = [{
...
+        const eligibleFunding = await getFunding(request) ?? undefined
         if (!eligibleFunding) {
           return h.view('no-response').takeover()
         }
         return h.view('funding/what-funding', new ViewModel(eligibleFunding, funding, error)).code(400).takeover()
       }
     },
     handler: async (request, h) => {
-      const { standard } = request.payload
+      const standards = typeof (request.payload.standard) === 'string' ? [request.payload.standard] : request.payload.standard
```
Comment (marctemp) [inline app/routes/funding/what-funding.js:43]: Knock-on effect from Joi not validating the standard payload is an array it has to explicitly check and cast if needs be

File: app/routes/funding/what-funding.js [hunk 4]
```diff
@@ -24,35 +27,37 @@ module.exports = [{
...
       const { data, agreement } = await cache.get(request)
 
-      const funding = data?.eligibleFunding.filter(x => standard.includes(x.code)) ?? []
+      const eligibleFunding = (data?.eligibleFunding ?? []).filter(fundingObj => standards.includes(fundingObj.code))
```
Comment (marctemp) [inline app/routes/funding/what-funding.js:46]: Variables funding renamed to eligibleFunding as there are 2 variables within the cache with these names and this line refers to the latter.

File: app/views/funding/what-funding.njk [hunk 5]
```diff
@@ -25,30 +24,55 @@
...
         <input type="hidden" name="crumb" value="{{crumb}}"/>
         {% if model.items.length > 0 %}
-          {{ govukRadios(model) }}
-          {{ govukButton({
-          text: "Continue",
-          attributes: {
-            id: "submit"
-          }
-        }) }}
+          {{ govukCheckboxes(model) }}
```
Comment (marctemp) [inline app/views/funding/what-funding.njk:36]: Mix of Nunjucks and HTML unfort here, could not find a way of mixing HTML elements with the Nunjucks component.

File: app/what-funding.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+const getChecked = (selected, standardItems) => {
+  if (Object.keys(standardItems).includes(selected)) {
+    standardItems[selected].checked = true
+  }
+  return standardItems
+}
+
+module.exports = { getEligible, getChecked, mapStandards }
```
Comment (marctemp) [inline app/what-funding.js:44]: Had to export for testing purposes, i.e. so the unit test could import them to call them - is there a way around this? Ideally if functions are only used internally to a module should they be visible, importing and callable to users from a different module?

File: app/routes/funding/what-funding.js (hunk 2 above)
Comment (johnwatson484) [reply app/routes/funding/what-funding.js:30]: I think if you use `required()` on the array it should fix it.

ie. `Joi.array().items(Joi.string()).single().required()`

File: app/routes/funding/models/what-funding.js (hunk 1 above)
Comment (johnwatson484) [reply app/routes/funding/models/what-funding.js:1]: Yes I agree.  Though I'd maybe tweak the split as there is some properties in the new module which are part of the nunjucks model such as `checked` and `hint`.

```
const standardDetails = {
  'sfi-moorland': {
    text: 'Funding for moorland or rough grazing',
    value: 'sfi-moorland',
    hint: { text: '£6.45 per hectare per year, plus a single payment of £148 per year.' },
    checked: false
  },
  'sfi-arable-soil': {
    text: 'Funding for arable and horticultural soil',
    value: 'sfi-arable-soil',
    hint: { text: '£22 to £40 per hectare per year based on what work you do.' },
    checked: false
  },
  'sfi-improved-grassland': {
    text: 'Funding for improved grassland soil',
    value: 'sfi-improved-grassland',
    hint: { text: '£28 to £58 per hectare per year based on what work you do.' },
    checked: false
  }
}
```

File: app/views/funding/what-funding.njk (hunk 5 above)
Comment (johnwatson484) [reply app/views/funding/what-funding.njk:36]: I think it' unavoidable in this case

File: app/funding/get-funding.js [hunk 7]
```diff
@@ -21,6 +21,7 @@ const requestEligibleFunding = async (organisation, callerId, correlationId) =>
   const { sbi, organisationId } = organisation
   const body = { sbi, callerId, organisationId }
 
+  console.log(`fund sbu: ${sbi} -- org ${organisationId} -- call ${callerId}`)
```
Comment (edi-defra) [inline app/funding/get-funding.js:24]: @marctemp I wonder if replacing this line with an actual logger might be better? 
LMK your thoughts on this.
Thanks

File: app/routes/funding/what-funding.js (hunk 2 above)
Comment (marctemp) [reply app/routes/funding/what-funding.js]: Yup, `Joi.array().items(Joi.string()).single().required()` worked, added and tests passing

File: app/routes/funding/models/what-funding.js (hunk 1 above)
Comment (marctemp) [reply app/routes/funding/models/what-funding.js:1]: Will address this when I change the agreement-calc to inc. moorland and grazing should look similar to:

`
.map(fundingObj => {
      return {
        text: fundingObj.name,
        value: fundingObj.code,
        hint: { text: fundingObj.hint },
        checked: isChecked(selected, fundingObj.code)
      }
`

as I alter the return from the agreement-calc to include these keys

File: app/routes/funding/models/what-funding.js [hunk 8]
```diff
@@ -9,12 +11,19 @@ function ViewModel (eligibleFunding, selected, error) {
...
+    items: mapStandards(eligibleFunding, selected).map(funding => { funding.checked = isChecked(selected, funding.value); return funding })
   }
 
   if (error) {
-    this.model.errorMessage = {
-      text: 'Please choose the funding you would like to apply for '
+    this.model.error = {
+      errorList: [
+        {
+          text: 'Select an option'
```
Comment (steve-dickinson) [inline app/routes/funding/models/what-funding.js:24]: Maybe change the text for the error to 

Please choose the funding you would like to apply for.

Also to make it consistent, add "There is a problem" title. 

```
this.model.error = {
      titleText: "There is a problem",
      errorList: [
        {
          text: 'Select an option'
```

**Review by steve-dickinson — DISMISSED** [review]
LGTM

**Review by steve-dickinson — APPROVED** [review]
LGTM

**Pull request merged**